himbaechel/uarch/gatemate/gatemate.cc

```diff
@@ -261,6 +263,45 @@ void GateMateImpl::postPlace()
 {
     repack();
     ctx->assignArchInfo();
+    for (auto &cell : ctx->cells) {
+        // We need to skip CPE_MULT since using CP outputs is mandatory
+        // even if output is actually not connected
+        bool marked_used = cell.second.get()->type == id_CPE_MULT;
+        // Can not use FF for OUT2 if CPE is used in bridge mode
+        if (cell.second.get()->type == id_CPE_FF && ctx->getBelLocation(cell.second.get()->bel).z == CPE_FF_U_Z)
+            marked_used = true;
+        if (marked_used)
+            used_cpes.emplace(ctx->getBelName(cell.second.get()->bel)[0]);
+    }
+}
+bool GateMateImpl::checkPipAvail(PipId pip) const
+{
+    const auto &extra_data =
+            *reinterpret_cast<const GateMatePipExtraDataPOD *>(chip_pip_info(ctx->chip_info, pip).extra_data.get());
+    if (extra_data.type != PipExtra::PIP_EXTRA_MUX || (extra_data.flags & MUX_ROUTING) == 0)
+        return true;
+    IdStringList id = ctx->getPipName(pip);
+    if (used_cpes.count(id[0]))
+        return false;
+    return true;
+}
+
+void GateMateImpl::notifyPipChange(PipId pip, NetInfo *net)
+{
+    const auto &extra_data = *reinterpret_cast<const GateMatePipExtraDataPOD *>(
+        chip_pip_info(ctx->chip_info, pip).extra_data.get());
+
+    if (extra_data.type != PipExtra::PIP_EXTRA_MUX || !(extra_data.flags & MUX_ROUTING))
+        return;
+
+    auto found = cpe_bridges.count(pip) > 0;
+
+    NPNR_ASSERT((!found && net != nullptr) || (found && net == nullptr));
+
+    if (found)
```

Comment:
if `cpe_bridges` is only used post-route it would be more efficient to only build it post-route, too (this is not really relevant for router2 but would be for router1 that binds pips as it goes)